### PR TITLE
Clean up interfaces

### PR DESCRIPTION
**Pull request by blink1073**
Puts the kernel interfaces in `IKernel` and `KernelMessage` namespaces, and puts the session interfaces in an `ISession` interface and marks everything as a JSON object.

Renames `NotebookSession -> Session`.

Temporarily add `json.ts` here until we switch to the new `phosphor` library.

Readmes and examples updated and verified.

### CODE REVIEW

**Comment by blink1073** [comment]
Probably best viewed as the new file content instead of a diff.

File: src/ikernel.ts [hunk 1]
```diff
@@ -14,1042 +14,1020 @@ import {
...
+} from './json';
 
 
 /**
- * The options object used to initialize a kernel.
+ * A namespace for kernel types, interfaces, and type checker functions.
  */
-export
-interface IKernelOptions {
+export namespace kernel {
```
Comment (afshin) [inline src/ikernel.ts]: Just a line break:
```
export
namespace kernel
```

**Comment by afshin** [comment]
Good use of the new JSON definitions, it makes these interfaces more accurate and type safe.

**Comment by blink1073** [comment]
Word.

**Comment by blink1073** [comment]
This is ready for review.  The substantial changes are in `ikernel.ts` and `isession.ts`, with the rest of the changes being the necessary modifications.

**Comment by afshin** [comment]
This is epic. It looks good, cleaner organization, better interfaces, and code conventions.

👍

**Pull request merged**